### PR TITLE
feat: persist app group and env group filter in local storage

### PR DESCRIPTION
**Pull request by arunjaindev**
# Description
feat: persist app group and env group filter in local storage

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/ApplicationGroup/AppGroupAppFilter.tsx [hunk 1]
```diff
@@ -87,11 +88,14 @@ export default function AppGroupAppFilter() {
         if (selectedFilterTab === AppFilterTabs.APP_FILTER) {
             setSelectedAppList(selectedValue)
             setSelectedGroupFilter([])
+            setFilterInLocalStorage(filterParentType, resourceId, selectedValue, [])
```
Comment (AbhishekA1509) [inline src/components/ApplicationGroup/AppGroupAppFilter.tsx:91]: NOTE: Please discuss this before moving ahead with this comment

Ideally this should be done through useStateFilters or there should be a mechanism to auto sync state with local storage, after discussion can take call.

File: src/components/ApplicationGroup/AppGroupAppFilter.tsx [hunk 2]
```diff
@@ -108,6 +112,51 @@ export default function AppGroupAppFilter() {
         appGroupFilterRef.current.onMenuOpen()
     }
 
+    useEffect(() => {
+        if (!appListOptions || !groupFilterOptions) {
+            return
+        }
+
+        const getAndSetItem = (localStorageKey: `${string}__filter`) => {
+            const localStorageValue = localStorage.getItem(localStorageKey)
```
Comment (AbhishekA1509) [inline src/components/ApplicationGroup/AppGroupAppFilter.tsx:121]: I believe there is some duplication with setFilterInLocalStorage

File: src/components/ApplicationGroup/AppGroupAppFilter.tsx [hunk 3]
```diff
@@ -108,6 +112,51 @@ export default function AppGroupAppFilter() {
...
+            if (!localStorageValue) {
+                return
+            }
+            const valueForCurrentResource = new Map(JSON.parse(localStorageValue)).get(resourceId)
```
Comment (AbhishekA1509) [inline src/components/ApplicationGroup/AppGroupAppFilter.tsx:125]: Please safely parse the local storage value

File: src/components/ApplicationGroup/AppGroupAppFilter.tsx [hunk 4]
```diff
@@ -108,6 +112,51 @@ export default function AppGroupAppFilter() {
...
+
+            if (filteredLocalStorageResourceList) {
+                setSelectedAppList(filteredLocalStorageResourceList)
+            }
+            if (filteredLocalStorageGroupList) {
+                setSelectedGroupFilter(filteredLocalStorageGroupList)
+            }
+        }
+
+        if (filterParentType === FilterParentType.app) {
```
Comment (AbhishekA1509) [inline src/components/ApplicationGroup/AppGroupAppFilter.tsx:153]: Please move the key into variable based on condition and then set the value

File: src/components/ApplicationGroup/AppGroupAppFilter.tsx [hunk 5]
```diff
@@ -108,6 +112,51 @@ export default function AppGroupAppFilter() {
...
+                return agg
+            }, {})
+
+            const groupFilterOptionsMap = groupFilterOptions.reduce<Record<string, true>>((agg, curr) => {
+                agg[curr.value] = true
+                return agg
+            }, {})
+
+            // filtering local storage lists acc to new appList/ envList or groupFilterList as local values might be deleted or does not exist anymore
+            const filteredLocalStorageResourceList = localStorageResourceList?.filter(({value}) => appListOptionsMap[value])
```
Comment (AbhishekA1509) [inline src/components/ApplicationGroup/AppGroupAppFilter.tsx:142]: Can add fallback into variable directly instead of null check here

**Review by AbhishekA1509 — CHANGES_REQUESTED** [review]

**Review by AbhishekA1509 — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**